arangod/Aql/QueryPlanCache.h

```diff
@@ -0,0 +1,216 @@
+// attribute name bind parameters. these bind parameters are relevant during the
+// query optimization phase, because they determine which collections and
+// indexes can be used by the query.
+// once an AQL query is started that has its "optimizePlanForCaching" flag
+// set to true, the Query object will generate a cache key object for the
+// query plan cache.
+// it will then perform a lookup in the query plan cache and run the query using
+// the serialized cached plan if there is any in the cache. if there is no plan
+// in the cache, the Query object will memorize its cache key and invoke the
+// query parser and optimizer. after the optimization, the Query object will
+// serialize the plan to velocypack and store the serialized plan in the plan
+// cache, using the already computed cache key.
+//
+// TODO:
+// - currently the cache is never invalidated, which is wrong. we must
+// invalidate
+//   the cache whenever
+//   - a collection gets dropped
+//   - maybe: the properties of a collection get changed (TODO: clarify if this
+//     can influence the query plans at all)
+//   - an index for a collection gets dropped
+//   - a new index for a collection gets created
+//   - a view gets dropped
+//   - the properties of an existing view get modified
+//   in the cluster, we need to track the changes of database objects which are
+//   executed via other coordinators by tailing the changes in the AgencyCache,
+//   and react accordingly. this is currently missing in this PR.
+// - dropping caches when a database is dropped is not necessary however,
+//   because each database has its own QueryPlanCache object
+// - the cache implementation is intentionally kept simple. each per-database
+//   cache object has a simple R/W lock, which may turn out to be a point of
+//   mutex contention at some point. in this case, we could shard each cache
+//   object so that it uses multiple buckets
+// - we currently disable caching whenever one of the following query options
+//   are used:
+//   - forceOneShardAttributeValue
+//   - inaccessibleCollections (Enterprise Edition only)
+//   - restrictToShards ("shardIds" query option)
+//   we need to figure out if there are additional query options that should
+//   disable the use of the plan cache, or if we can lift the restrictions on
+//   the above attributes. especially the "forceOneShardAttributeValue" is used
+//   in reality, and we may want to support it.
+// - we need to ensure that permissions for collections/views are still verified
+//   even for queries that come from the plan cache. this is required because a
+//   privileged user's query may insert the query into the plan cache, and an
+//   unprivileged user's query may reuse the plan cache entry, even though the
+//   unprivileged user should not have access to the collections/views used in
+//   the query. we can copy the permissions checks from the query results cache
+//   that we already have.
```

Comment:
I checked that an at least in the single server case, I get a forbidden error, although the cache was used to load the plan. It appears also in the code, that the collection permissions are checked _after_ the optimization.

Reply:
Permission checks have been added.